borsh-rs/borsh/src/de/mod.rs

```diff
@@ -167,9 +271,67 @@ where
     T: BorshDeserialize,
 {
     #[inline]
-    fn deserialize<R: Read>(reader: &mut R) -> Result<Self, Error> {
+    fn deserialize<R: BufRead>(reader: &mut R) -> Result<Self, Error> {
         let len = u32::deserialize(reader)?;
-        if size_of::<T>() == 0 {
+        if len == 0 {
+            Ok(Vec::new())
+        } else if T::use_unsafe_transmute() {
+            let mut result = Vec::with_capacity(hint::cautious::<u8>(len));
+            let mut len = len as usize;
+            while len > 0 {
+                let read_len = {
+                    let buf = reader.fill_buf()?;
+                    if buf.is_empty() {
+                        return Err(std::io::Error::new(
+                            std::io::ErrorKind::InvalidInput,
+                            ERROR_UNEXPECTED_LENGTH_OF_INPUT,
+                        ));
+                    }
+                    if buf.len() > len {
+                        result.extend_from_slice(&buf[..len]);
+                        len
+                    } else {
+                        result.extend_from_slice(&buf);
+                        buf.len()
+                    }
+                };
+                len -= read_len;
+                reader.consume(read_len)
+            }
+            let result = unsafe {
```

Comment:
I don't understand why is it needed. It takes vector `result` and constructs exactly the same vector from its memory.

Reply:
`result` is `Vec<u8>` while we need `Vec<T>`

We can add an additional check for `size_of::<T>() == size_of::<u8>()`